File: TRENDX/Networking/TrendXAPIClient.swift
//
//  TrendXAPIClient.swift
//  TRENDX
//

import Foundation

enum TrendXAPIError: LocalizedError {
    case notConfigured
    case invalidResponse
    case server(status: Int, message: String)

    var errorDescription: String? {
        switch self {
        case .notConfigured:
            return "TRENDX API is not configured for this build."
        case .invalidResponse:
            return "The API returned an invalid response."
        case let .server(status, message):
            return "TRENDX API request failed (\(status)): \(message)"
        }
    }
}

final class TrendXAPIClient {
    let config: TrendXAPIConfig

    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    init(config: TrendXAPIConfig = .current, session: URLSession = .shared) {
        self.config = config
        self.session = session

        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let value = try container.decode(String.self)
            if let date = ISO8601DateFormatter.trendxFractional.date(from: value)
                ?? ISO8601DateFormatter.trendxInternet.date(from: value) {
                return date
            }
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Invalid ISO-8601 date: \(value)"
            )
        }
        self.decoder = decoder

        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        encoder.dateEncodingStrategy = .iso8601
        self.encoder = encoder
    }

    func get<T: Decodable>(_ path: String, accessToken: String? = nil) async throws -> T {
        try await request(path, method: "GET", accessToken: accessToken, bodyData: nil)
    }

    func post<T: Decodable, Body: Encodable>(
        _ path: String,
        accessToken: String? = nil,
        body: Body
    ) async throws -> T {
        let data = try encoder.encode(body)
        return try await request(path, method: "POST", accessToken: accessToken, bodyData: data)
    }

    private func request<T: Decodable>(
        _ path: String,
        method: String,
        accessToken: String?,
        bodyData: Data?
    ) async throws -> T {
        guard let baseURL = config.baseURL, config.isConfigured else {
            throw TrendXAPIError.notConfigured
        }

        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw TrendXAPIError.invalidResponse
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = bodyData

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw TrendXAPIError.invalidResponse
        }

        guard (200..<300).contains(http.statusCode) else {
            let message = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw TrendXAPIError.server(status: http.statusCode, message: message)
        }

        if T.self == EmptyResponse.self {
            return EmptyResponse() as! T
        }
        return try decoder.decode(T.self, from: data)
    }
}

struct EmptyResponse: Decodable {}

private extension ISO8601DateFormatter {
    static let trendxInternet: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()

    static let trendxFractional: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()
}

File: android/app/src/main/java/com/trendx/app/networking/TrendXAPIClient.kt
package com.trendx.app.networking

import android.util.Log
import com.trendx.app.config.TrendXAPIConfig
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.HttpResponseValidator
import io.ktor.client.plugins.ResponseException
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.defaultRequest
import io.ktor.client.plugins.logging.LogLevel
import io.ktor.client.plugins.logging.Logger as KtorLogger
import io.ktor.client.plugins.logging.Logging
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNamingStrategy

// Equivalent of TRENDX/Networking/TrendXAPIClient.swift. Uses kotlinx.serialization
// with the snake_case naming strategy so backend payloads decode straight into
// the camelCase Kotlin data classes — matching the iOS client's
// `convertFromSnakeCase` / `convertToSnakeCase` JSON config.

sealed class TrendXAPIError(message: String) : RuntimeException(message) {
    object NotConfigured : TrendXAPIError("TRENDX API is not configured for this build.")
    object InvalidResponse : TrendXAPIError("The API returned an invalid response.")
    class Server(val status: Int, val payload: String) :
        TrendXAPIError("TRENDX API request failed ($status): $payload")
}

class TrendXAPIClient(
    val config: TrendXAPIConfig = TrendXAPIConfig.current
) {
    @OptIn(kotlinx.serialization.ExperimentalSerializationApi::class)
    val json: Json = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
        encodeDefaults = true
        coerceInputValues = true
        namingStrategy = JsonNamingStrategy.SnakeCase
    }

    // @PublishedApi internal so the inline `get` / `post` below can call
    // through. Logcat tag "TrendXAPI" — `adb logcat -s TrendXAPI` to see
    // every request body, response status, and any non-2xx payload.
    @PublishedApi
    internal val client: HttpClient = HttpClient(OkHttp) {
        expectSuccess = false
        install(ContentNegotiation) { json(this@TrendXAPIClient.json) }
        install(Logging) {
            logger = object : KtorLogger {
                // Log.i so the messages survive Logcat's default INFO+
                // filter — Log.d gets hidden unless you opt in to "Verbose".
                override fun log(message: String) { Log.i("TrendXAPI", message) }
            }
            // ALL = method + URL + headers + body for both request and
            // response. Verbose, but exactly what we need while we're
            // diagnosing why /bootstrap doesn't bring back real data.
            level = LogLevel.ALL
        }
        defaultRequest {
            contentType(ContentType.Application.Json)
        }
        HttpResponseValidator {
            validateResponse { response ->
                val status = response.status.value
                if (status !in 200..299) {
                    val body = runCatching { response.bodyAsText() }.getOrDefault("")
                    Log.w("TrendXAPI", "HTTP $status from ${response.call.request.url}: ${body.take(280)}")
                    throw TrendXAPIError.Server(status, body)
                }
            }
        }
    }

    suspend inline fun <reified T> get(path: String, accessToken: String? = null): T {
        val url = absoluteUrl(path)
        return try {
            client.get(url) {
                accessToken?.let { header(HttpHeaders.Authorization, "Bearer $it") }
            }.body()
        } catch (e: ResponseException) {
            throw TrendXAPIError.Server(e.response.status.value, e.message ?: "")
        }
    }

    suspend inline fun <reified T, reified Body : Any> post(
        path: String,
        body: Body,
        accessToken: String? = null
    ): T {
        val url = absoluteUrl(path)
        return try {
            client.post(url) {
                accessToken?.let { header(HttpHeaders.Authorization, "Bearer $it") }
                setBody(body)
            }.body()
        } catch (e: ResponseException) {
            throw TrendXAPIError.Server(e.response.status.value, e.message ?: "")
        }
    }

    fun absoluteUrl(path: String): String {
        // Plain string concat. Earlier we used Ktor's URLBuilder, which
        // appended an extra path segment to "https://host" (its starting
        // path was already an empty segment), producing "https://host//path"
        // and silent 404s on Railway. String math matches iOS's
        // URL(string:relativeTo:) behavior without the surprise.
        val base = config.baseUrl?.trimEnd('/') ?: throw TrendXAPIError.NotConfigured
        val normalized = if (path.startsWith("/")) path else "/$path"
        return base + normalized
    }
}
